### PR TITLE
MGMT-11156: Cluster validation, merge the print and validate methods

### PR DESCRIPTION
**Pull request by paul-maidment**
Same as https://github.com/openshift/assisted-service/commit/4e04230522992aad4d93551b8d935aa0d8b2ed0b but for cluster validations instead of host validations

The present implementation of cluster validation has a problem, it splits the validation into two phases. The validation and the printing of the validation message.

This leads to a situation in some cases where it is not possible to use the output of a validation to determine the message that should be sent back to the user, there are some workarounds but these involve poor practices such as code repetition.

This commit changes the validation to be a single method. Due to the complexity of the refactor, no effort has been made to make each of the validation methods their "most efficient" it is assumed that these fixes will take place in subsequent commits.

Because the validation methods made heavy use of return statements, the style has had to be changed to an if/else style and some code has had to be reorganised a little.

In theory, the existing tests should be sufficient to act as a "smoke" test for this functionality provided that there is enough coverage.

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @tsorya 
/cc @ori-amizur 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Review by gamli75 — APPROVED** [review]

File: internal/cluster/cluster_suite_test.go [hunk 1]
```diff
@@ -11,6 +12,7 @@ import (
 func TestCluster(t *testing.T) {
 	RegisterFailHandler(Fail)
 	common.InitializeDBTest()
+	time.Sleep(10 * time.Second)
```
Comment (nmagnezi) [inline internal/cluster/cluster_suite_test.go:15]: Is this intentional? or just for debugging purposes?

**Review by nmagnezi — COMMENTED** [review]
The `hold` label is intentional?

File: internal/cluster/cluster_suite_test.go (hunk 1 above)
Comment (paul-maidment) [reply internal/cluster/cluster_suite_test.go]: Just debugging, removed it now

Comment (paul-maidment) [reply internal/cluster/cluster_suite_test.go]: and unheld it as at the time of writing, the hold was intentional but no longer necessary

File: Dockerfile.assisted-service-build [hunk 2]
```diff
@@ -28,4 +28,4 @@ COPY ./hack/setup_env.sh ./dev-requirements.txt ./
 RUN ./setup_env.sh podman_remote && ./setup_env.sh assisted_service
 RUN chmod g+xw -R $GOPATH && chmod g+xw -R $(go env GOROOT)
 COPY ./data /tmp/data
-COPY . .
\ No newline at end of file
+COPY . .
```
Comment (omertuc) [inline Dockerfile.assisted-service-build:31]: ?

**Comment by omertuc** [comment]
The commit message would be much more helpful if it were just "Same as 4e04230522992aad4d93551b8d935aa0d8b2ed0b but for cluster validations instead of host validations"

This way I don't have to think "Does this change make sense" but instead I can think "This was already done & approved for host validations, of course it makes sense we should also do it for cluster validations"

File: internal/cluster/validator.go [hunk 3]
```diff
@@ -473,38 +355,22 @@ func isReadyToInstall(status string) bool {
 	return funk.ContainsString(allowedStatuses, status)
 }
 
-func (v *clusterValidator) allHostsAreReadyToInstall(c *clusterPreprocessContext) ValidationStatus {
+func (v *clusterValidator) allHostsAreReadyToInstall(c *clusterPreprocessContext) (ValidationStatus, string) {
 	readyToInstall := true
 	for _, host := range c.cluster.Hosts {
 		readyToInstall = readyToInstall && isReadyToInstall(swag.StringValue(host.Status))
 	}
```
Comment (carbonin) [inline internal/cluster/validator.go:362]:
```suggestion
	readyToInstall := true
	for _, host := range c.cluster.Hosts {
		if !isReadyToInstall(swag.StringValue(host.Status)) {
		        readyToInstall = false
		        break
		}
	}
```

I know this is unchanged, but it seems strange to loop through all of the hosts if you already know one isn't ready.

File: internal/cluster/validator.go [hunk 4]
```diff
@@ -528,177 +379,127 @@ func checkCidrsOverlapping(cluster *common.Cluster) error {
...
+	serviceCidrDefined, _ = v.isServiceCidrDefined(c)
+
+	//If one of the required Cidr fields is empty return Pending status
+	if !validationStatusToBool(clusterCidrDefined) || !validationStatusToBool(serviceCidrDefined) {
+		status = ValidationPending
+	} else {
+		machineCidrDefined, _ := v.isMachineCidrDefined(c)
+		machineNetworkRequired := network.IsMachineNetworkRequired(c.cluster)
+		if machineNetworkRequired && !validationStatusToBool(machineCidrDefined) {
+			status = ValidationPending
```
Comment (carbonin) [inline internal/cluster/validator.go:420]: I think you could probably combine this with the message handling and return early, no?

`network.IsMachineNetworkRequired` is checking user managed networking so it'd imagine it's more accurate than the check below.

File: internal/cluster/validator.go [hunk 5]
```diff
@@ -528,177 +379,127 @@ func checkCidrsOverlapping(cluster *common.Cluster) error {
...
-		return "No ntp problems found"
-	case ValidationFailure:
-		return fmt.Sprintf("Hosts' clocks are not synchronized (there's more than a %d minutes gap between clocks), "+
-			"please configure an NTP server via DHCP or set clocks manually.", common.MaximumAllowedTimeDiffMinutes)
-	default:
-		return fmt.Sprintf("Unexpected status %s", status)
+	if synced {
+		return ValidationSuccess, "No ntp problems found"
 	}
+	return ValidationFailure, fmt.Sprintf("Hosts' clocks are not synchronized (there's more than a %d minutes gap between clocks), "+
```
Comment (carbonin) [inline internal/cluster/validator.go:503]: Small nit, but whenever possible I think the success case should be the least indented. So I'd reverse this.
```suggestion
	if !synced {
		return ValidationFailure, fmt.Sprintf("Hosts' clocks are not synchronized (there's more than a %d minutes gap between clocks), "+
	}
	return ValidationSuccess, "No ntp problems found"
```

File: internal/cluster/cluster_suite_test.go (hunk 1 above)
Comment (carbonin) [inline internal/cluster/cluster_suite_test.go:15]: Please remove this

File: internal/cluster/validator.go [hunk 6]
```diff
@@ -402,58 +402,50 @@ func (v *clusterValidator) noCidrsOverlapping(c *clusterPreprocessContext) (Vali
...
+		v.log.WithError(err).Errorf("Getting service address families for cluster %s", c.cluster.ID.String())
+		return ValidationError, "Bad CIDR(s) appears in one of the networks"
+	}
+	clusterNetworkFamilies, err := network.CidrsToAddressFamilies(network.GetClusterNetworkCidrs(c.cluster))
+	if err != nil {
+		v.log.WithError(err).Errorf("Getting cluster address families for cluster %s", c.cluster.ID.String())
+		return ValidationError, "Bad CIDR(s) appears in one of the networks"
+	}
+	// serviceNetworkFamilies should only ever have a maximum of two distinct families.
+	// clusterNetworkFamilies may have multiple indistinct families so need to be canonized prior to comparison.
```
Comment (carbonin) [inline internal/cluster/validator.go:432]: Something I was confused about here was what "canonize" meant.
Maybe that's worth explaining too?

**Review by carbonin — CHANGES_REQUESTED** [review]
New test cases look good, but the names and structure could be better

**Review by carbonin — APPROVED** [review]

**Pull request merged**